File: src/leetcode_grind/day_306.rs
// https://leetcode.com/problems/reconstruct-itinerary/
pub fn find_itinerary(tickets: Vec<Vec<String>>) -> Vec<String> {
    use std::collections::HashMap;
    let mut adj = HashMap::new();
    for ticket in &tickets {
        let (from, to) = (ticket[0].clone(), ticket[1].clone());
        adj.entry(from).or_insert(vec![]).push(to);
    }
    let mut visited = HashMap::new();
    for (k, v) in &mut adj {
        v.sort();
        visited.entry(k.clone()).or_insert(vec![false; v.len()]);
    }

    let mut res = vec![];
    let mut path = vec!["JFK".to_string()];

    fn backtrack(
        from: String,
        adj: &HashMap<String, Vec<String>>,
        visited: &mut HashMap<String, Vec<bool>>,
        path: &mut Vec<String>,
        res: &mut Vec<String>,
        tickets: usize,
    ) -> bool {
        if path.len() == tickets + 1 {
            *res = path.clone();
            return true;
        }
        if !adj.contains_key(&from) {
            return false;
        }
        for (i, to) in adj[&from].iter().enumerate() {
            if !visited[&from][i] {
                visited.get_mut(&from).unwrap()[i] = true;
                path.push(to.clone());
                let res = backtrack(to.clone(), adj, visited, path, res, tickets);
                path.pop();
                visited.get_mut(&from).unwrap()[i] = false;
                if res {
                    return true;
                }
            }
        }
        false
    }

    backtrack(
        "JFK".to_string(),
        &adj,
        &mut visited,
        &mut path,
        &mut res,
        tickets.len(),
    );

    res
}

pub fn find_itinerary_ii(tickets: Vec<Vec<String>>) -> Vec<String> {
    use std::collections::HashMap;
    let mut adj = HashMap::new();
    for ticket in tickets {
        let (from, to) = (ticket[0].clone(), ticket[1].clone());
        adj.entry(from).or_insert(vec![]).push(to);
    }

    fn dfs(
        from: String,
        adj: &HashMap<String, Vec<String>>,
        visited: &mut HashMap<String, usize>,
        path: &mut Vec<String>,
    ) {
        if adj.contains_key(&from) {
            while visited[&from] < adj.get(&from).unwrap().len() {
                let to = adj[&from][visited[&from]].clone();
                *visited.get_mut(&from).unwrap() += 1;
                dfs(to, adj, visited, path);
            }
        }

        path.push(from);
    }

    let mut path = vec![];
    let mut visited = HashMap::new();
    for (k, v) in &mut adj {
        visited.insert(k.clone(), 0);
        v.sort();
    }

    dfs("JFK".to_owned(), &adj, &mut visited, &mut path);
    path.reverse();
    path
}

File: java/grind/Day306.java
package grind;
import java.util.*;

class Solution {
    public List<String> findItinerary(List<List<String>> tickets) {
        var adj = new HashMap<String, List<String>>();
        for (var ticket : tickets) {
            adj.putIfAbsent(ticket.get(0), new ArrayList<>(0));
            adj.get(ticket.get(0)).add(ticket.get(1));
        }
        adj.values().forEach(Collections::sort);
        var result = new ArrayList<String>();
        var path = new ArrayList<String>() {
            {
                add("JFK");
            }
        };
        var visited = new HashSet<String>();
        var dfs = new Object() {
            boolean dfs(String from) {
                if (path.size() == tickets.size() + 1) {
                    result.addAll((List<String>) path.clone());
                    return true;
                }
                if (!adj.containsKey(from)) {
                    return false;
                }
                for (var i = 0; i < adj.get(from).size(); i++) {
                    var key = String.format("%s->%s", from, i);
                    if (!visited.contains(key)) {
                        visited.add(key);
                        var to = adj.get(from).get(i);
                        path.add(to);
                        var res = dfs(to);
                        path.remove(path.size() - 1);
                        visited.remove(key);
                        if (res) {
                            return true;
                        }
                    }
                }
                return false;
            }
        };

        dfs.dfs("JFK");
        return result;
    }

    public List<String> findItineraryHirtzholder(List<List<String>> tickets) {
        var adj = new HashMap<String, List<String>>();
        for (var ticket : tickets) {
            adj.putIfAbsent(ticket.get(0), new ArrayList<>());
            adj.get(ticket.get(0)).add(ticket.get(1));
        }

        adj.values().forEach(Collections::sort);

        var path = new ArrayList<String>();
        var visited = new HashSet<String>();

        var dfs = new Object() {
            void dfs(String from) {
                if (adj.containsKey(from)) {
                    for (var i = 0; i < adj.get(from).size(); i++) {
                        var key = String.format("%s->%s", from, i);
                        if (!visited.contains(key)) {
                            visited.add(key);
                            var to = adj.get(from).get(i);
                            dfs(to);
                        }

                    }
                }
                path.add(from);
            }
        };

        dfs.dfs("JFK");
        Collections.reverse(path);
        return path;
    }
}